sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4

```diff
@@ -594,16 +594,13 @@ qualifiedName
     : identifier ('.' identifier)*
     ;
 
-// Identifier that also allows the use of a number of SQL keywords (mainly for backwards compatibility).
-looseIdentifier
-    : identifier
-    | FROM
-    | TO
-    | TABLE
-    | WITH
+identifier
+    : strictIdentifier
+    | ANTI | FULL | INNER | LEFT | SEMI | RIGHT | NATURAL | UNION | INTERSECT | EXCEPT | DATABASE | ON | JOIN | CROSS
+    | SELECT | FROM | WHERE | HAVING | FROM | TO | TABLE | WITH | NOT
```

Comment:
The semantic is actually quite simple: any keyword that can be use after a table alias must be added to the `strictIdentifier` rule. In practice this means that we need to add all JOIN and SET-operator keywords. I have updated the description.

I am not sure how lenient database systems are. I like the parser to be as lenient as possible, since we also use the table identifier and data type parsing in a lot of other places in the codebase.